Sources/Articles/Documentation.docc/2024/beginners-guide-to-protocol-buffers-and-grpc-with-swift/beginners-guide-to-protocol-buffers-and-grpc-with-swift.md

```diff
@@ -0,0 +1,360 @@
+
+Protocol Buffers are extensively used across various applications, such as network communication, data storage, and configuration management. They offer a language-neutral, platform-neutral, and extensible approach to serializing structured data.
+
+## What is gRPC?
+
+[gRPC](https://grpc.io/) (gRPC Remote Procedure Call) is an open-source framework developed by Google. It enables efficient communication between distributed systems and services, allowing them to collaborate seamlessly.
+
+It is quite similar to the OpenAPI standard, but Protocol Buffers are more efficient, as they use a binary format instead of JSON. 
+
+When using the OpenAPI format, schemas are defined using a JSON or YAML file. In contrast, Protocol Buffers define the schema of data structures using a `.proto` file, which specifies the structure and data types to be serialized.
+
+
+## The Protocol Buffer Compiler
+
+The protocol buffer compiler (protoc) is a tool that generates source code for data serialization and gRPC communication from `.proto` schema files.
+
+The protocol buffers compiler supports many [languages](https://protobuf.dev/getting-started/), including Swift. To [install](https://grpc.io/docs/protoc-installation/) protoc on your machine, follow the instructions based on your operating system:
+
+```sh
+# install on macOS 
+brew install protobuf
+
+# install on Linux
+apt install -y protobuf-compiler
+
+# manual install, on any operating system
+PB_REL="https://github.com/protocolbuffers/protobuf/releases"
+curl -LO $PB_REL/download/v25.1/protoc-25.1-linux-x86_64.zip
+unzip protoc-25.1-linux-x86_64.zip -d $HOME/.local
+export PATH="$PATH:$HOME/.local/bin"
+```
+
+The `protoc` command is now ready to use. The next step involves defining the data structure for a sample todo application. This is accomplished by creating a `todo.proto` file, which will outline the models for the application. Below is the complete protobuf definition:
+
+```protobuf
+// todo_messages.proto
+syntax = "proto3";
+
+// 1.
+package todos;
+
+// 2.
+message Empty {}
+
+// 3.
+message TodoList {
+    repeated Todo todos = 1;
+}
+
+// 4.
+message TodoID {
+    string todoID = 1;
+}
+
+// 5.
+message Todo {
+    optional string todoID = 1;
+    string title = 2;
+    bool completed = 3;
+}
+```
+
+1. The package `todos` defines the namespace for this protobuf file.
+2. The `Empty` message is an empty structure, often used for simple requests or responses without data.
+3. The `TodoList` message contains a list of `Todo` items, marked as a repeated field.
+4. The `TodoID` message holds a single string field to represent the ID of a specific Todo item.
+5. The `Todo` message defines a task with an optional `todoID`, a `title`, and a `completed` status represented by a boolean.
```

Comment:
```suggestion
1. Define a `todos` namespace for this protobuf file.
2. Create the `Empty` message with an empty structure - which contains no data.
3. The `TodoID` message holds a single string field to represent the ID of a specific "Todo" item.
4. A `Todo` message defines a task with an optional `todoID`, a `title`, and a `completed` status represented by a boolean.
5. Finally, the `TodoList` message contains a list of `Todo` items, through a repeated field.
```

I suggest moving TodoList below the Todo and TodoID here, since we can then see dependencies in order